tests/Microsoft.Identity.Test.Integration.net45/HeadlessTests/AuthorityTests.cs

```diff
@@ -47,5 +54,67 @@ public async Task AuthorityMigrationAsync()
 
             Assert.IsNotNull(authResult.AccessToken);
         }
+
+        [TestMethod]
+        public async Task FailedAuthorityValidationTestAsync()
+        {
+            var labResponse = await LabUserHelper.GetDefaultUserAsync().ConfigureAwait(false);
+            var user = labResponse.User;
+
+            IPublicClientApplication pca = PublicClientApplicationBuilder
+                .Create(labResponse.AppId)
+                .WithAuthority("https://bogus.microsoft.com/common")
+                .Build();
+
+            Trace.WriteLine("Acquire a token using a not so common authority alias");
+
+            MsalServiceException exception = await AssertException.TaskThrowsAsync<MsalServiceException>(() =>
+                 pca.AcquireTokenByUsernamePassword(
+                    s_scopes,
+                     user.Upn,
+                     new NetworkCredential("", user.GetOrFetchPassword()).SecurePassword)
+                     .ExecuteAsync())
+                .ConfigureAwait(false);
+
+            Assert.IsTrue(exception.Message.Contains("AADSTS50049"));
+            Assert.AreEqual("invalid_instance", exception.ErrorCode);
+        }
+
+        /// <summary>
+        /// If this test fails, please update the <see cref="KnownMetadataProvider"/> to 
+        /// use whatever Evo uses (i.e. the aliases, preferred network / metadata from the url below).
+        /// </summary>
+        [TestMethod]
+        public async Task KnownInstanceMetadataIsUpToDateAsync()
+        {
+            string validDiscoveryUri = @"https://login.microsoftonline.com/common/discovery/instance?api-version=1.1&authorization_endpoint=https%3A%2F%2Flogin.microsoftonline.com%2Fcommon%2Foauth2%2Fv2.0%2Fauthorize";
+            HttpClient httpClient = new HttpClient();
+            HttpResponseMessage discoveryResponse = await httpClient.SendAsync(
+                new HttpRequestMessage(
+                    HttpMethod.Get,
+                    validDiscoveryUri)).ConfigureAwait(false);
+            string discoveryJson = await discoveryResponse.Content.ReadAsStringAsync().ConfigureAwait(false);
+
+            var actualMetadata = JsonHelper.DeserializeFromJson<InstanceDiscoveryResponse>(discoveryJson).Metadata;
+            var expectedMetadata = KnownMetadataProvider.GetAllEntriesForTest().Values;
+
+            Enumerable.SequenceEqual(actualMetadata, expectedMetadata, new InstanceDiscoveryMetadataEntryComparer());
```

Comment:
> Enumerable.SequenceEqual [](start = 11, length = 25)

this is cool

Reply:
Yes, I was initially trying to use CollectionAssert which does the same, but it is a very old function that doesn't support typed collections, so it's a bit ugly. This is much more elegant, but I see that I have a bug - I don't actually assert on the result of this function :)